src/Microsoft.Diagnostics.Runtime/ClrValue.cs

```diff
@@ -0,0 +1,1077 @@
+        /// Returns the address of the value.
+        /// </summary>
+        public abstract ulong Address { get; }
+
+        /// <summary>
+        /// Returns the result of ReadPointer(Address) if this value is an object.
+        /// </summary>
+        protected abstract ulong Object { get; }
+
+        /// <summary>
+        /// Returns whether Address is an interior value.
+        /// </summary>
+        public abstract bool Interior { get; }
+
+        /// <summary>
+        /// Returns whether this value is null (or otherwise cannot be accessed).
+        /// </summary>
+        public virtual bool IsNull { get { return Address == 0 || (ClrRuntime.IsObjectReference(ElementType) && Object == 0); } }
+
+        /// <summary>
+        /// Returns the element type of this value.
+        /// </summary>
+        public abstract ClrElementType ElementType { get; }
+        
+        /// <summary>
+        /// The runtime associated with this value.
+        /// </summary>
+        public virtual ClrRuntime Runtime { get { return _runtime; } }
+
+        /// <summary>
+        /// Returns the type of this value.
+        /// </summary>
+        public virtual ClrType Type
+        {
+            get
+            {
+                HeapBase heap = _runtime.HeapBase;
+                ClrElementType el = ElementType;
+
+                if (ClrRuntime.IsPrimitive(el))
+                    return heap.GetBasicType(el);
+
+                if (ClrRuntime.IsObjectReference(el))
+                    return AsObject().Type;
+
+                return GetStructElementType();
+            }
+        }
+
+        /// <summary>
+        /// Obtains the element type when ElementType is a struct.
+        /// </summary>
+        /// <returns>The ClrType of this value, or ClrHeap.ErrorType if it could not be obtained.</returns>
+        protected abstract ClrType GetStructElementType();
+
+        #region Converters
+
+        /// <summary>
+        /// Returns a ClrObject for this value.
+        /// </summary>
+        /// <returns>A ClrObject for this value.  If the value is null, then ClrObject.IsNull will the true and ClrObject.Type
+        /// will equal ClrHeap.ErrorType.</returns>
+        public virtual ClrObject AsObject()
+        {
+            if (!ClrRuntime.IsObjectReference(ElementType))
+                throw new InvalidOperationException("Value is not an object.");
+
+            if (IsNull)
+                throw new NullReferenceException();
+
+            ClrHeap heap = Runtime.GetHeap();
+
+            ulong obj = Object;
+            return new ClrObject(obj, obj != 0 ? heap.GetObjectType(obj) : heap.NullType);
+        }
+
+        /// <summary>
+        /// Converts the value to a boolean.  Note that this method is a conversion, not a cast.  Similarly sized data will be converted.
+        /// </summary>
+        /// <returns>This value as a boolean.</returns>
+        public virtual bool AsBoolean()
+        {
```

Comment:
C# does support overloading the implicit or explicit casting operators. Rather writing:
bool b = value.AsBoolean()

Customers could write:
bool b = (bool)value;
OR:
bool b = value;